keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -447,17 +447,35 @@ It replaces its internal store with the collected items (syncWith) and reuses th
 
 #### API changes
 
-Extend the optional `ResourceVersionMatch` query parameter of `ListOptions` with the following enumeration value:
+Extend the `ListOptions` struct with the following field:
 
 ```
-const (
-    // ResourceVersionMatchMostRecent matches data at the most recent ResourceVersion.
-    // The returned data is consistent, that is, served from etcd via a quorum read.
-    // For watch calls, it begins with synthetic "Added" events of all resources up to the most recent ResourceVersion.
-    // It ends with a synthetic "Bookmark" event containing the most recent ResourceVersion.
-    // For list calls, it has the same semantics as leaving ResourceVersion and ResourceVersionMatch unset.
-    ResourceVersionMatchMostRecent ResourceVersionMatch = "MostRecent"
-)
+type ListOptions struct {
+    ...
+
+    // SendInitialEvents, when set together with Watch option,
+    // being the watch stream with synthetic init events to build the
+    // whole state of all resources followed by a synthetic "Bookmark"
+    // event containing a ResourceVersion after which the server
+    // continues streaming events.
+    //
+    // When SendInitialEvents option is set, we require ResourceVersionMatch
+    // option to also be set. The semantic of the watch request is as following:
+    // - ResourceVersionMatch = NotOlderThan
+    //   It starts with sending initial events for all objects (at some resource
+    //   version), potentially followed by an event stream until the state
+    //   becomes synced to a resource version as fresh as the one provided by
+    //   the ResourceVersion option. At this point, a synthetic bookmark event
+    //   is send and watch stream is continued to be send.
+    //   If RV is unset, this is interpreted as "consistent read" and the
+    //   bookmark event is send when the state is synced to the moment when
+    //   request started being processed.
+    // - ResourceVersionMath = Exact
+    //   Unsupported error is returned.
+    // - ResourceVersionMatch unset (or set to any other value)
+    //   BadRequest error is returned.
+    SendInitialEvents bool
+}
```

Comment:
If we would agree that this is the right API, then the proposal needs to be updated in couple other places, but I won't do that, until we agree for the API change.

Reply:
What happens to the existing way of requesting a watch which sends the existing entire collection first?

Reply:
I was going to say option 2, but since people have to start doing something extra to pay attention to the bookmarks anyway, maybe it's OK to deprecate it?

Reply:
thinking about how the current behavior of watching from rv="" or rv="0" interacts with the behavior controlled by the existing `resourceVersionMatch` and `allowWatchBookmarks` options, and the proposed sendInitialEvents option...

it would be _really_ nice if rv="" / rv="0" was less of a special case

* It always sends initial events, which implicitly matches the behavior of `sendInitialEvents=true`.
  * It feels silly to not allow `?watch=true&resourceVersion=0&sendInitialEvents=true&...` when we're actually sending initial events anyway.
  * it also feels silly to not allow `?watch=true&resourceVersion=0&sendInitialEvents=false&...`, when that would let us do less work and would let someone avoid receiving all existing items as `Added` events if they really just wanted to start a watch stream and didn't care what version they started at
* > It doesn't send any bookmark.
  * It could if `allowWatchBookmarks=true` was sent... that seems useful for someone wanting to know when the initial set of Added events was done
* > It's partially consistent across etcd3 and watchcache
  * Since we don't advertise whether watch cache is on for a given resource type, and we don't allow use of `resourceVersionMatch` with watch currently, I don't see how callers could rely on the "current state" etcd3 semantic as opposed to the "last synced state" watch cache semantic.
  * weirdly, we [document different semantics](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch) for watching from rv=0 and rv="", but the watch cache doesn't appear to distinguish them: https://github.com/kubernetes/kubernetes/blob/07801383afe46fde2a62086da1f46266cd393462/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L482-L488
  * I think it would be reasonable for `resourceVersion=0` and `resourceVersion=` to imply `resourceVersionMatch=NotOlderThan` and `sendInitialEvents=true`

I don't see much benefit to deprecating and trying to remove watching with rv=0 (not that we could realistically remove that anyway as a GA behavior), especially if we're contemplating letting watchers explicitly request to stream all initial events anyway.

Reply:
Thanks for your thoughts Jordan.
I think I agree with most of the above - just adding something thoughts here:

> It always sends initial events, which implicitly matches the behavior of sendInitialEvents=true

That makes perfect sense to me.
My only concern here is that effectively mean that:
- if RV="" or RV="0", sendInitialEvents is somewhat defaulted to true
- otherwise sendInitialEvents is default to false

OTOH, it's what it is now, so if you're ok with that - I'm fine with it too.

> It could if allowWatchBookmarks=true was sent...

Agree.

> weirdly, we document different semantics for watching from rv=0 and rv="", but the watch cache doesn't appear to distinguish them:

Hmm - I must admit that I completely missed that.
It's a bug that is independent from this proposal - I just filed: https://github.com/kubernetes/kubernetes/issues/114845


> I think it would be reasonable for resourceVersion=0 and resourceVersion= to imply resourceVersionMatch=NotOlderThan and sendInitialEvents=true

SGTM


Let me update this proposal to reflect that better (will try to do that later today or early next week).

Reply:
I update the proposal to reflect that.